compiler/rustc_codegen_ssa/src/base.rs

```diff
@@ -840,20 +840,30 @@ impl CrateInfo {
         //
         // In order to get this left-to-right dependency ordering, we use the reverse
         // postorder of all crates putting the leaves at the right-most positions.
-        let used_crates = tcx
+        let mut compiler_builtins = None;
+        let mut used_crates: Vec<_> = tcx
             .postorder_cnums(())
             .iter()
             .rev()
             .copied()
-            .filter(|&cnum| !tcx.dep_kind(cnum).macros_only())
+            .filter(|&cnum| {
+                let link = !tcx.dep_kind(cnum).macros_only();
+                if link && tcx.is_compiler_builtins(cnum) {
+                    compiler_builtins = Some(cnum);
```

Comment:
Shouldn't compiler builtins be included even if it isn't used according to name resolution?

Reply:
`postorder_cnums` returns all dependency crates.
If compiler builtins end up only being used as a proc macro dependency somehow (`macros_only`), then they shouldn't get into `used_crates` and be linked.
`CrateInfo::compiler_builtins` field is also only used twice - in `fn ignored_for_lto` and `fn add_upstream_rust_crates`, which are not for proc macro dependencies.

Reply:
I thought compiler_builtins was automatically injected for all crates except libcore and I assumed that it wouldn't be listed as regular dependency. I can't find the relevant code though, so maybe I was wrong.